File: java/eshop-cqrs-ddd-microservices-excercise/common/src/main/java/com/marekturis/common/infrastructure/messaging/RabbitMQConsumer.java
package com.marekturis.common.infrastructure.messaging;

import com.marekturis.common.domain.event.EventHandler;
import com.marekturis.common.domain.event.JsonParsableEvent;
import com.marekturis.common.domain.event.ParsableEvent;
import com.rabbitmq.client.AMQP;
import com.rabbitmq.client.Channel;
import com.rabbitmq.client.DefaultConsumer;
import com.rabbitmq.client.Envelope;

import java.io.IOException;

/**
 * @author Marek Turis
 */
public class RabbitMQConsumer extends DefaultConsumer {

	EventHandler handler;

	public RabbitMQConsumer(Channel channel, EventHandler eventHandler) {
		super(channel);
		handler = eventHandler;
	}

	@Override
	public void handleDelivery(String consumerTag, Envelope envelope,
							   AMQP.BasicProperties properties, byte[] body) throws IOException {
		String jsonMessage = new String(body, "UTF-8");
		ParsableEvent parsableEvent = new JsonParsableEvent(jsonMessage);
		try {
			handler.handle(parsableEvent);
		}
		catch (Exception ex) {
			System.out.println("Handling of event was unsuccesful");
			ex.printStackTrace();
		}
		finally {
			getChannel().basicAck(envelope.getDeliveryTag(), false);
		}
	}
}

File: dotnet/eshop-cqrs-ddd-microservices-excercise/Com.Marekturis.Common/Infrastructure/Messaging/RabbitMQConsumer.cs
using Com.Marekturis.Common.Domain.Event;
using RabbitMQ.Client;
using EventHandler = Com.Marekturis.Common.Domain.Event.EventHandler;

namespace Com.Marekturis.Common.Infrastructure.Messaging
{
    public class RabbitMQConsumer : DefaultBasicConsumer
    {
        private readonly IModel channel;
        private readonly EventHandler eventHandler;

        public RabbitMQConsumer(EventHandler eventHandler, IModel channel)
        {
            this.eventHandler = eventHandler;
            this.channel = channel;
        }

        public override void HandleBasicDeliver(string consumerTag, ulong deliveryTag, bool redelivered, string exchange, string routingKey,
            IBasicProperties properties, byte[] body)
        {
            var jsonMessage = System.Text.Encoding.UTF8.GetString(body, 0, body.Length);
            ParsableEvent parsableEvent = new JsonParsableEvent(jsonMessage);
            try {
                eventHandler.Handle(parsableEvent);
            } finally {
                channel.BasicAck(deliveryTag, false);
            }
        }
    }
}